packages/salesforcedx-vscode-apex/src/commands/apexActionController.ts

```diff
@@ -132,4 +130,127 @@ export class ApexActionController {
   }
+
+  private saveOasAsErsMetadata = async (oasSpec: string, fullPath: string): Promise<void> => {
+    // Replace the schema section in the ESR file if it already exists
+    let existingContent;
+    if (fs.existsSync(fullPath)) {
+      existingContent = fs.readFileSync(fullPath, 'utf8');
+    }
+    const namedCredential = await this.showNamedCredentialsQuickPick();
+
+    const updatedContent = existingContent
+      ? existingContent.replace(/<schema>([\s\S]*?)<\/schema>/, `<schema>${oasSpec.replaceAll('"', '&apos;')}</schema>`)
+      : [
+          '<?xml version="1.0" encoding="UTF-8"?>',
+          '<ExternalServiceRegistration xmlns="http://soap.sforce.com/2006/04/metadata">',
+          `\t<description>${path.basename(fullPath).split('.')[0]} External Service</description>`,
+          `\t<label>${path.basename(fullPath).split('.')[0]}</label>`,
+          `\t<namedCredentialReference>${namedCredential}</namedCredentialReference>`,
+          '\t<registrationProviderType>Custom</registrationProviderType>',
+          `\t<schema>${oasSpec.replaceAll('"', '&apos;')}</schema>`,
+          '\t<schemaType>OpenApi3</schemaType>',
+          '\t<schemaUploadFileExtension>yaml</schemaUploadFileExtension>',
+          `\t<schemaUploadFileName>${path.basename(fullPath).split('.')[0].toLowerCase()}_openapi</schemaUploadFileName>`,
+          '\t<status>Complete</status>',
+          '\t<systemVersion>5</systemVersion>',
+          '</ExternalServiceRegistration>'
+        ].join('\n');
+    try {
+      // Step 3: Write File
+      fs.writeFileSync(fullPath, updatedContent);
+      await vscode.workspace.openTextDocument(fullPath).then((newDocument: vscode.TextDocument) => {
+        void vscode.window.showTextDocument(newDocument);
+      });
+    } catch (error) {
+      throw new Error(nls.localize('artifact_failed', error.message));
+    }
+  };
+
+  private pathExists = async (filename: string): Promise<string> => {
+    // Step 1: Prompt for Folder
+    const folder = await this.getFolderForArtifact();
+    if (!folder) {
+      throw new Error(nls.localize('no_folder_selected'));
+    }
+
+    // Step 2: Verify folder exists and if not create it
+    if (!fs.existsSync(folder)) {
+      fs.mkdirSync(folder);
+    }
+
+    // Step 2: Check if File Exists
+    const fullPath = path.join(folder, filename);
+    if (fs.existsSync(fullPath)) {
+      const shouldOverwrite = await this.confirmOverwrite();
+      if (!shouldOverwrite) {
+        throw new Error(nls.localize('operation_cancelled'));
+      }
+    }
+    return fullPath;
+  };
+
+  private confirmOverwrite = async (): Promise<boolean> => {
+    const response = await vscode.window.showWarningMessage(
+      nls.localize('file_exists'),
+      { modal: true },
+      nls.localize('overwrite'),
+      nls.localize('cancel')
+    );
+    return response === nls.localize('overwrite');
+  };
+
+  private getFolderForArtifact = async (): Promise<string | undefined> => {
```

Comment:
Is it not possible to get the default location for ESR from SDR?

Reply:
I think we can using [src/registry/metadataRegistry.json](https://github.com/forcedotcom/source-deploy-retrieve/blob/7a75847af83777a2673165c909672857c32f90e8/src/registry/metadataRegistry.json#L273) ? I'll look at it! Thanks!